File: src/knowledge/search/result-ranker.ts
/**
 * ResultRanker — 多维融合排序器
 *
 * 将不同搜索策略的结果按多维度评分进行融合排序：
 * - TF-IDF 文本相关性
 * - 符号匹配精确度
 * - 上下文位置权重（definition > import > usage > comment）
 * - 依赖图中心度
 * - 文件新鲜度（recency）
 * - 文件类型权重
 *
 * 每个维度的权重可配置。
 */

import { IndexSearchHit } from '../indexing/inverted-index';
import { SymbolSearchHit } from '../indexing/symbol-index';
import { DependencyGraph } from '../indexing/dependency-graph';
import { MinHeap } from '../utils/min-heap';

// ============================================================================
// 类型定义
// ============================================================================

/** 融合后的排序结果 */
export interface RankedResult {
  filePath: string;
  /** 最终综合得分（0-1） */
  finalScore: number;
  /** 各维度得分明细 */
  breakdown: ScoreDimensions;
  /** 来源标记 */
  sources: ('index' | 'symbol' | 'dependency')[];
}

/** 各维度得分 */
export interface ScoreDimensions {
  /** 倒排索引 TF-IDF 得分 */
  tfidf: number;
  /** 符号匹配得分 */
  symbolMatch: number;
  /** 上下文位置权重 */
  positionWeight: number;
  /** 依赖图中心度 */
  centrality: number;
  /** 文件新鲜度（最近修改的文件得分更高） */
  recency: number;
  /** 文件类型权重 */
  typeWeight: number;
  /** 最近编辑加权（0-1） */
  recentEditBoost: number;
  /** 当前 scope 加权（0-1） */
  scopeBoost: number;
}

/** 排序权重配置 */
export interface RankWeights {
  tfidf: number;
  symbolMatch: number;
  positionWeight: number;
  centrality: number;
  recency: number;
  typeWeight: number;
}

/** 文件时间信息（由调用方传入） */
export interface FileTimestamps {
  /** 文件路径 → 最后修改时间戳(ms) */
  get(filePath: string): number | undefined;
}

/** 额外排序信号（与基础 rank 权重解耦） */
export interface RankBoostSignals {
  /** 当前任务聚焦目录/文件 */
  preferredScopes?: string[];
  /** 最近编辑的文件集合 */
  recentEditedFiles?: Set<string>;
}

/** 聚合条目（消除 any 类型） */
interface FileScoreEntry {
  sources: Set<string>;
  tfidf: number;
  symbolMatch: number;
  positionWeight: number;
  centrality: number;
  recency: number;
  typeWeight: number;
  recentEditBoost: number;
  scopeBoost: number;
}

/** 默认权重 */
const DEFAULT_WEIGHTS: RankWeights = {
  tfidf: 0.30,
  symbolMatch: 0.28,
  positionWeight: 0.15,
  centrality: 0.10,
  recency: 0.07,
  typeWeight: 0.10,
};

// ============================================================================
// ResultRanker 类
// ============================================================================

export class ResultRanker {
  private weights: RankWeights;

  constructor(weights?: Partial<RankWeights>) {
    this.weights = { ...DEFAULT_WEIGHTS, ...weights };
    // 归一化权重
    const total = Object.values(this.weights).reduce((sum, w) => sum + w, 0);
    if (total > 0) {
      for (const key of Object.keys(this.weights) as Array<keyof RankWeights>) {
        this.weights[key] /= total;
      }
    }
  }

  /**
   * 融合多源搜索结果并排序
   * @param fileTimestamps 文件最后修改时间映射（用于 recency 计算）
   * @param weightOverrides 可选权重覆盖（用于查询意图感知）
   */
  rank(
    indexHits: IndexSearchHit[],
    symbolHits: SymbolSearchHit[],
    dependencyGraph: DependencyGraph | null,
    maxResults = 20,
    fileTimestamps?: FileTimestamps,
    weightOverrides?: Partial<RankWeights>,
    boostSignals?: RankBoostSignals
  ): RankedResult[] {
    // 按需合并权重覆盖并归一化
    let effectiveWeights = this.weights;
    if (weightOverrides) {
      const merged = { ...this.weights, ...weightOverrides };
      const total = Object.values(merged).reduce((sum, w) => sum + w, 0);
      if (total > 0) {
        effectiveWeights = { ...merged };
        for (const key of Object.keys(effectiveWeights) as Array<keyof RankWeights>) {
          effectiveWeights[key] /= total;
        }
      }
    }
    // 按文件聚合所有来源的分数
    const fileMap = new Map<string, FileScoreEntry>();

    // 1. 处理倒排索引结果
    for (const hit of indexHits) {
      const entry = this.getOrCreate(fileMap, hit.filePath);
      entry.sources.add('index');
      entry.tfidf = Math.max(entry.tfidf, hit.score);
      entry.positionWeight = Math.max(
        entry.positionWeight,
        this.contextToWeight(hit.bestContext)
      );
    }

    // 2. 处理符号匹配结果
    for (const hit of symbolHits) {
      const entry = this.getOrCreate(fileMap, hit.symbol.filePath);
      entry.sources.add('symbol');
      entry.symbolMatch = Math.max(entry.symbolMatch, hit.score);
      // 导出符号额外加分
      if (hit.symbol.isExported) {
        entry.typeWeight = Math.max(entry.typeWeight, 0.3);
      }
    }

    // 3. 补充依赖图信息
    if (dependencyGraph?.isReady) {
      for (const [filePath, entry] of fileMap.entries()) {
        entry.centrality = dependencyGraph.getCentrality(filePath);
      }
    }

    // 4. 补充文件新鲜度（recency）
    if (fileTimestamps) {
      const now = Date.now();
      for (const [filePath, entry] of fileMap.entries()) {
        const mtime = fileTimestamps.get(filePath);
        if (mtime) {
          entry.recency = this.calculateRecency(mtime, now);
        }
      }
    }

    // 5. 额外信号加权：最近编辑 + 当前 scope
    if (boostSignals?.recentEditedFiles && boostSignals.recentEditedFiles.size > 0) {
      for (const [filePath, entry] of fileMap.entries()) {
        if (boostSignals.recentEditedFiles.has(filePath)) {
          entry.recentEditBoost = 1;
        }
      }
    }
    const preferredScopes = boostSignals?.preferredScopes || [];
    if (preferredScopes.length > 0) {
      for (const [filePath, entry] of fileMap.entries()) {
        entry.scopeBoost = this.calculateScopeBoost(filePath, preferredScopes);
      }
    }

    // 优化 #17: MinHeap Top-K 替换全量 sort+slice
    const heap = new MinHeap<RankedResult>(maxResults, (a, b) => a.finalScore - b.finalScore);

    for (const [filePath, entry] of fileMap.entries()) {
      const breakdown: ScoreDimensions = {
        tfidf: entry.tfidf,
        symbolMatch: entry.symbolMatch,
        positionWeight: entry.positionWeight,
        centrality: entry.centrality,
        recency: entry.recency,
        typeWeight: entry.typeWeight,
        recentEditBoost: entry.recentEditBoost,
        scopeBoost: entry.scopeBoost,
      };

      let finalScore =
        breakdown.tfidf * effectiveWeights.tfidf +
        breakdown.symbolMatch * effectiveWeights.symbolMatch +
        breakdown.positionWeight * effectiveWeights.positionWeight +
        breakdown.centrality * effectiveWeights.centrality +
        breakdown.recency * effectiveWeights.recency +
        breakdown.typeWeight * effectiveWeights.typeWeight;

      const scopedBonus = breakdown.scopeBoost * 0.15;
      const recentEditBonus = breakdown.recentEditBoost * 0.10;
      finalScore *= (1 + scopedBonus + recentEditBonus);

      // 优化 #7: 多来源交叉加分
      // 同时被 2 个来源命中 → +10%，3 个来源 → +20%
      if (entry.sources.size >= 3) {
        finalScore *= 1.20;
      } else if (entry.sources.size >= 2) {
        finalScore *= 1.10;
      }

      heap.push({
        filePath,
        finalScore,
        breakdown,
        sources: Array.from(entry.sources) as RankedResult['sources'],
      });
    }

    return heap.toSortedDescArray();
  }

  // ==========================================================================
  // 私有方法
  // ==========================================================================

  private getOrCreate(
    map: Map<string, FileScoreEntry>,
    filePath: string
  ): FileScoreEntry {
    let entry = map.get(filePath);
    if (!entry) {
      entry = {
        sources: new Set<string>(),
        tfidf: 0,
        symbolMatch: 0,
        positionWeight: 0,
        centrality: 0,
        recency: 0,
        typeWeight: 0,
        recentEditBoost: 0,
        scopeBoost: 0,
      };
      map.set(filePath, entry);
    }
    return entry;
  }

  /**
   * scope 加权：
   * - 文件精确命中 scope 路径：1.0
   * - 文件位于 scope 子路径：0.8
   * - 文件路径包含 scope 片段：0.4
   */
  private calculateScopeBoost(filePath: string, preferredScopes: string[]): number {
    const normalizedFilePath = filePath.replace(/\\/g, '/').toLowerCase();
    let best = 0;

    for (const scope of preferredScopes) {
      const normalizedScope = scope.replace(/\\/g, '/').replace(/^\.\/+/, '').toLowerCase();
      if (!normalizedScope) continue;

      if (normalizedFilePath === normalizedScope) {
        best = Math.max(best, 1.0);
        continue;
      }
      if (normalizedFilePath.startsWith(`${normalizedScope}/`)) {
        best = Math.max(best, 0.8);
        continue;
      }
      if (normalizedFilePath.includes(`/${normalizedScope}/`) || normalizedFilePath.endsWith(`/${normalizedScope}`)) {
        best = Math.max(best, 0.4);
      }
    }

    return best;
  }

  /**
   * 计算文件新鲜度得分（0-1）
   * 优化 #8: 指数衰减函数替代离散阶梯
   * 半衰期 = 72 小时（3 天），即 3 天前的文件得分约 0.5
   */
  private calculateRecency(lastModified: number, now: number): number {
    const ageHours = (now - lastModified) / (1000 * 60 * 60);
    const halfLife = 72; // 半衰期：72 小时
    return Math.pow(0.5, ageHours / halfLife);
  }

  /**
   * 上下文类型 → 权重
   */
  private contextToWeight(context: string): number {
    switch (context) {
      case 'definition': return 1.0;
      case 'import': return 0.6;
      case 'usage': return 0.4;
      case 'string': return 0.2;
      case 'comment': return 0.2;
      default: return 0.3;
    }
  }
}

File: crates/magi-knowledge-store/src/result_ranker.rs
use std::collections::{HashMap, HashSet};

use serde::{Deserialize, Serialize};

use crate::code_tokenizer::TokenContext;
use crate::inverted_index::IndexSearchHit;
use crate::min_heap::MinHeap;
use crate::symbol_index::SymbolSearchHit;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct RankWeights {
    pub tfidf: f64,
    pub symbol_match: f64,
    pub position_weight: f64,
    pub centrality: f64,
    pub recency: f64,
    pub type_weight: f64,
}

impl Default for RankWeights {
    fn default() -> Self {
        Self {
            tfidf: 0.30,
            symbol_match: 0.28,
            position_weight: 0.15,
            centrality: 0.10,
            recency: 0.07,
            type_weight: 0.10,
        }
    }
}

impl RankWeights {
    fn normalized(&self) -> Self {
        let total = self.tfidf
            + self.symbol_match
            + self.position_weight
            + self.centrality
            + self.recency
            + self.type_weight;
        if total <= 0.0 {
            return self.clone();
        }
        Self {
            tfidf: self.tfidf / total,
            symbol_match: self.symbol_match / total,
            position_weight: self.position_weight / total,
            centrality: self.centrality / total,
            recency: self.recency / total,
            type_weight: self.type_weight / total,
        }
    }

    fn merge_overrides(&self, overrides: &RankWeights) -> Self {
        Self {
            tfidf: if overrides.tfidf > 0.0 {
                overrides.tfidf
            } else {
                self.tfidf
            },
            symbol_match: if overrides.symbol_match > 0.0 {
                overrides.symbol_match
            } else {
                self.symbol_match
            },
            position_weight: if overrides.position_weight > 0.0 {
                overrides.position_weight
            } else {
                self.position_weight
            },
            centrality: if overrides.centrality > 0.0 {
                overrides.centrality
            } else {
                self.centrality
            },
            recency: if overrides.recency > 0.0 {
                overrides.recency
            } else {
                self.recency
            },
            type_weight: if overrides.type_weight > 0.0 {
                overrides.type_weight
            } else {
                self.type_weight
            },
        }
    }
}

#[derive(Clone, Debug)]
pub struct ScoreDimensions {
    pub tfidf: f64,
    pub symbol_match: f64,
    pub position_weight: f64,
    pub centrality: f64,
    pub recency: f64,
    pub type_weight: f64,
    pub recent_edit_boost: f64,
    pub scope_boost: f64,
}

#[derive(Clone, Debug)]
pub struct RankedResult {
    pub file_path: String,
    pub final_score: f64,
    pub breakdown: ScoreDimensions,
    pub sources: Vec<String>,
}

#[derive(Clone, Debug, Default)]
pub struct RankBoostSignals {
    pub preferred_scopes: Vec<String>,
    pub recent_edited_files: HashSet<String>,
}

struct FileScoreEntry {
    sources: HashSet<String>,
    tfidf: f64,
    symbol_match: f64,
    position_weight: f64,
    centrality: f64,
    recency: f64,
    type_weight: f64,
    recent_edit_boost: f64,
    scope_boost: f64,
}

impl Default for FileScoreEntry {
    fn default() -> Self {
        Self {
            sources: HashSet::new(),
            tfidf: 0.0,
            symbol_match: 0.0,
            position_weight: 0.0,
            centrality: 0.0,
            recency: 0.0,
            type_weight: 0.0,
            recent_edit_boost: 0.0,
            scope_boost: 0.0,
        }
    }
}

pub struct ResultRanker {
    weights: RankWeights,
}

pub struct RankInput<'a> {
    pub index_hits: &'a [IndexSearchHit],
    pub symbol_hits: &'a [SymbolSearchHit],
    pub centrality_map: &'a HashMap<String, f64>,
    pub max_results: usize,
    pub file_timestamps: &'a HashMap<String, u64>,
    pub weight_overrides: Option<&'a RankWeights>,
    pub boost_signals: Option<&'a RankBoostSignals>,
}

impl Default for ResultRanker {
    fn default() -> Self {
        Self::new(None)
    }
}

impl ResultRanker {
    pub fn new(weights: Option<RankWeights>) -> Self {
        let w = weights.unwrap_or_default().normalized();
        Self { weights: w }
    }

    pub fn rank(&self, input: RankInput<'_>) -> Vec<RankedResult> {
        let effective_weights = match input.weight_overrides {
            Some(overrides) => self.weights.merge_overrides(overrides).normalized(),
            None => self.weights.clone(),
        };

        let mut file_map: HashMap<String, FileScoreEntry> = HashMap::new();

        for hit in input.index_hits {
            let entry = file_map.entry(hit.file_path.clone()).or_default();
            entry.sources.insert("index".to_string());
            entry.tfidf = entry.tfidf.max(hit.score);
            entry.position_weight = entry
                .position_weight
                .max(context_to_weight(hit.best_context));
        }

        for hit in input.symbol_hits {
            let entry = file_map.entry(hit.symbol.file_path.clone()).or_default();
            entry.sources.insert("symbol".to_string());
            entry.symbol_match = entry.symbol_match.max(hit.score);
            if hit.symbol.is_exported {
                entry.type_weight = entry.type_weight.max(0.3);
            }
        }

        for (file_path, entry) in file_map.iter_mut() {
            if let Some(&c) = input.centrality_map.get(file_path) {
                entry.centrality = c;
            }
        }

        let now = now_millis();
        for (file_path, entry) in file_map.iter_mut() {
            if let Some(&mtime) = input.file_timestamps.get(file_path) {
                entry.recency = calculate_recency(mtime, now);
            }
        }

        if let Some(signals) = input.boost_signals {
            if !signals.recent_edited_files.is_empty() {
                for (file_path, entry) in file_map.iter_mut() {
                    if signals.recent_edited_files.contains(file_path) {
                        entry.recent_edit_boost = 1.0;
                    }
                }
            }
            if !signals.preferred_scopes.is_empty() {
                for (file_path, entry) in file_map.iter_mut() {
                    entry.scope_boost = calculate_scope_boost(file_path, &signals.preferred_scopes);
                }
            }
        }

        let cmp = |a: &RankedResult, b: &RankedResult| -> std::cmp::Ordering {
            a.final_score
                .partial_cmp(&b.final_score)
                .unwrap_or(std::cmp::Ordering::Equal)
                .then_with(|| b.file_path.cmp(&a.file_path))
        };
        let mut heap = MinHeap::new(input.max_results, cmp);

        for (file_path, entry) in file_map {
            let breakdown = ScoreDimensions {
                tfidf: entry.tfidf,
                symbol_match: entry.symbol_match,
                position_weight: entry.position_weight,
                centrality: entry.centrality,
                recency: entry.recency,
                type_weight: entry.type_weight,
                recent_edit_boost: entry.recent_edit_boost,
                scope_boost: entry.scope_boost,
            };

            let mut final_score = breakdown.tfidf * effective_weights.tfidf
                + breakdown.symbol_match * effective_weights.symbol_match
                + breakdown.position_weight * effective_weights.position_weight
                + breakdown.centrality * effective_weights.centrality
                + breakdown.recency * effective_weights.recency
                + breakdown.type_weight * effective_weights.type_weight;

            let scope_bonus = breakdown.scope_boost * 0.15;
            let recent_edit_bonus = breakdown.recent_edit_boost * 0.10;
            final_score *= 1.0 + scope_bonus + recent_edit_bonus;

            if entry.sources.len() >= 3 {
                final_score *= 1.20;
            } else if entry.sources.len() >= 2 {
                final_score *= 1.10;
            }

            heap.push(RankedResult {
                file_path,
                final_score,
                breakdown,
                sources: entry.sources.into_iter().collect(),
            });
        }

        heap.into_sorted_desc()
    }
}

fn context_to_weight(ctx: TokenContext) -> f64 {
    match ctx {
        TokenContext::Definition => 1.0,
        TokenContext::Import => 0.6,
        TokenContext::Usage => 0.4,
        TokenContext::String => 0.2,
        TokenContext::Comment => 0.2,
    }
}

fn calculate_recency(last_modified: u64, now: u64) -> f64 {
    let age_hours = (now.saturating_sub(last_modified)) as f64 / (1000.0 * 60.0 * 60.0);
    let half_life = 72.0;
    0.5_f64.powf(age_hours / half_life)
}

fn calculate_scope_boost(file_path: &str, preferred_scopes: &[String]) -> f64 {
    let norm_fp = file_path.replace('\\', "/").to_lowercase();
    let mut best = 0.0_f64;

    for scope in preferred_scopes {
        let norm_scope = scope
            .replace('\\', "/")
            .trim_start_matches("./")
            .to_lowercase();
        if norm_scope.is_empty() {
            continue;
        }

        if norm_fp == norm_scope {
            best = best.max(1.0);
        } else if norm_fp.starts_with(&format!("{}/", norm_scope)) {
            best = best.max(0.8);
        } else if norm_fp.contains(&format!("/{}/", norm_scope))
            || norm_fp.ends_with(&format!("/{}", norm_scope))
        {
            best = best.max(0.4);
        }
    }

    best
}

fn now_millis() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::symbol_index::{MatchType, SymbolEntry, SymbolKind};

    #[test]
    fn test_weight_normalization() {
        let w = RankWeights::default().normalized();
        let total =
            w.tfidf + w.symbol_match + w.position_weight + w.centrality + w.recency + w.type_weight;
        assert!((total - 1.0).abs() < 0.001);
    }

    #[test]
    fn test_rank_basic() {
        let ranker = ResultRanker::new(None);
        let index_hits = vec![IndexSearchHit {
            file_path: "src/auth.rs".to_string(),
            score: 5.0,
            hit_lines: vec![10, 20],
            matched_tokens: vec!["auth".to_string()],
            best_context: TokenContext::Definition,
        }];
        let symbol_hits = vec![SymbolSearchHit {
            symbol: SymbolEntry {
                name: "authenticate".to_string(),
                kind: SymbolKind::Function,
                file_path: "src/auth.rs".to_string(),
                line: 10,
                end_line: Some(30),
                is_exported: true,
                container: None,
                signature: None,
            },
            score: 80.0,
            match_type: MatchType::Exact,
        }];
        let results = ranker.rank(RankInput {
            index_hits: &index_hits,
            symbol_hits: &symbol_hits,
            centrality_map: &HashMap::new(),
            max_results: 10,
            file_timestamps: &HashMap::new(),
            weight_overrides: None,
            boost_signals: None,
        });
        assert_eq!(results.len(), 1);
        assert!(results[0].final_score > 0.0);
        assert!(results[0].sources.len() >= 2);
    }

    #[test]
    fn test_cross_source_bonus() {
        let ranker = ResultRanker::new(None);
        let index_hits = vec![IndexSearchHit {
            file_path: "a.rs".to_string(),
            score: 1.0,
            hit_lines: vec![1],
            matched_tokens: vec!["x".to_string()],
            best_context: TokenContext::Usage,
        }];
        let symbol_hits = vec![SymbolSearchHit {
            symbol: SymbolEntry {
                name: "x".to_string(),
                kind: SymbolKind::Function,
                file_path: "a.rs".to_string(),
                line: 1,
                end_line: None,
                is_exported: false,
                container: None,
                signature: None,
            },
            score: 1.0,
            match_type: MatchType::Exact,
        }];
        let mut centrality = HashMap::new();
        centrality.insert("a.rs".to_string(), 0.5);
        let r2 = ranker.rank(RankInput {
            index_hits: &index_hits,
            symbol_hits: &symbol_hits,
            centrality_map: &centrality,
            max_results: 10,
            file_timestamps: &HashMap::new(),
            weight_overrides: None,
            boost_signals: None,
        });
        assert!(r2[0].sources.len() >= 2);
    }

    #[test]
    fn test_recency() {
        let now = now_millis();
        let recent = calculate_recency(now - 1000, now);
        let old = calculate_recency(now - 72 * 3600 * 1000, now);
        assert!(recent > old);
        assert!((old - 0.5).abs() < 0.05);
    }

    #[test]
    fn test_scope_boost() {
        assert_eq!(
            calculate_scope_boost("src/auth.rs", &["src/auth.rs".into()]),
            1.0
        );
        assert_eq!(
            calculate_scope_boost("src/auth/login.rs", &["src/auth".into()]),
            0.8
        );
        assert!((calculate_scope_boost("lib/src/auth/x.rs", &["auth".into()]) - 0.4).abs() < 0.01);
    }
}
